### PR TITLE
Refactored FhirPath engine to work against IScopedNode

### PR DESCRIPTION
**Pull request by Kasdejong**
## Description
The FhirPath engine now works against IScopedNode

## Related issues
#2915 

## Testing
All existing FP tests are running except for a few cases of polymorphism. These are ignored for now in wait of refactoring against Base.

## Breaking changes
Refactored most signatures in the internals of the FP engine to use IScopedNode instead of ITypedElement. Surface API remains unchanged (through extension methods on ITE and Base), apart from changes to the FhirEvaluationContext fields.

### CODE REVIEW

File: src/Hl7.Fhir.Base/FhirPath/Expressions/Invokee.cs [hunk 1]
```diff
@@ -186,16 +186,16 @@ public static Invokee WrapLogic(Func<Func<bool?>, Func<bool?>, bool?> func)
...
+        public static Invokee Return(IScopedNode value)
         {
-            return (_, __) => (new[] { (ITypedElement)value });
+            return (_, _) => [value];
         }
 
-        public static Invokee Return(IEnumerable<ITypedElement> value)
+        public static Invokee Return(IEnumerable<IScopedNode> value)
         {
             return (_, __) => value;
```
Comment (Kasdejong) [inline src/Hl7.Fhir.Base/FhirPath/Expressions/Invokee.cs]: be consistent :)

File: src/Hl7.Fhir.Base/FhirPath/Expressions/SymbolTableInit.cs [hunk 2]
```diff
@@ -396,29 +398,30 @@ IEnumerable<ITypedElement> runForeach()
...
+                    IEnumerable<IScopedNode> newFocus = [element];
                     var newContext = ctx.Nest(newFocus);
                     newContext.SetThis(newFocus);
-                    newContext.SetIndex(ElementNode.CreateList(index));
+                    newContext.SetIndex([new Integer(index)]);
                     index++;
 
                     var candidates = lambda(newContext, InvokeeFactory.EmptyArgs);
-                    var uniqeNewNodes = candidates.Except(fullResult, EqualityOperators.TypedElementEqualityComparer);
+                    var uniqeNewNodes = candidates.Except(fullResult, EqualityOperators.TypedElementEqualityComparer).Select(ite => ite.ToScopedNode());
```
Comment (Kasdejong) [inline src/Hl7.Fhir.Base/FhirPath/Expressions/SymbolTableInit.cs]: use ToScopednodes()

File: src/Hl7.Fhir.Base/FhirPath/Expressions/Typecasts.cs [hunk 3]
```diff
@@ -77,7 +81,7 @@ private static Cast getImplicitCast(object f, Type to)
 
             if (typeof(P.Any).IsAssignableFrom(to) && !fromElemList)
             {
-                if (f is ITypedElement te && te.InstanceType == "Quantity") return o => ParseQuantity((ITypedElement)o);
+                if (f is IScopedNode te && te.InstanceType == "Quantity") return o => ParseQuantity((IScopedNode)o);
```
Comment (Kasdejong) [inline src/Hl7.Fhir.Base/FhirPath/Expressions/Typecasts.cs]: Replace instanceType with Type

File: src/Hl7.Fhir.Base/FhirPath/Functions/CollectionOperators.cs [hunk 4]
```diff
@@ -30,44 +32,47 @@ internal static class CollectionOperators
...
+        public static bool Not(this IEnumerable<IScopedNode> focus)
             => focus.Count() > 1
             ? throw Error.InvalidOperation($"Operator {nameof(Not)} is not applicable for collections with more than one item.")
             : !focus.BooleanEval().Value;
 
-        public static IEnumerable<ITypedElement> DistinctUnion(this IEnumerable<ITypedElement> a, IEnumerable<ITypedElement> b)
-            => a.Union(b, EqualityOperators.TypedElementEqualityComparer);
+        [TemporarilyChanged] // We cast all of them to scoped nodes for now. This will not be necessary once we define a clear Equality operator for IScopedNode
+        public static IEnumerable<IScopedNode> DistinctUnion(this IEnumerable<IScopedNode> a, IEnumerable<IScopedNode> b)
+            => a.Union(b, EqualityOperators.TypedElementEqualityComparer).Select(ite => ite.ToScopedNode());
```
Comment (Kasdejong) [inline src/Hl7.Fhir.Base/FhirPath/Functions/CollectionOperators.cs]: use ToScopedNodes()

File: src/Hl7.Fhir.Base/FhirPath/Functions/CollectionOperators.cs [hunk 5]
```diff
@@ -30,44 +32,47 @@ internal static class CollectionOperators
...
 
-        public static bool Contains(this IEnumerable<ITypedElement> focus, ITypedElement value)
+        public static bool Contains(this IEnumerable<IScopedNode> focus, IScopedNode value)
             => focus.Contains(value, EqualityOperators.TypedElementEqualityComparer);
 
-        public static IEnumerable<ITypedElement> Distinct(this IEnumerable<ITypedElement> focus)
-            => focus.Distinct(EqualityOperators.TypedElementEqualityComparer);
+        [TemporarilyChanged] // We cast all of them to scoped nodes for now. This will not be necessary once we define a clear Equality operator for IScopedNode
+        public static IEnumerable<IScopedNode> Distinct(this IEnumerable<IScopedNode> focus)
+            => focus.Distinct(EqualityOperators.TypedElementEqualityComparer).Select(ite => ite.ToScopedNode());
```
Comment (Kasdejong) [inline src/Hl7.Fhir.Base/FhirPath/Functions/CollectionOperators.cs]: use ToScopedNodes()

File: src/Hl7.Fhir.Base/FhirPath/Functions/CollectionOperators.cs [hunk 6]
```diff
@@ -30,44 +32,47 @@ internal static class CollectionOperators
...
 
-        public static bool SubsetOf(this IEnumerable<ITypedElement> focus, IEnumerable<ITypedElement> other)
+        public static bool SubsetOf(this IEnumerable<IScopedNode> focus, IEnumerable<IScopedNode> other)
             => focus.All(fitem => other.Contains(fitem));
 
-        public static IEnumerable<ITypedElement> Intersect(this IEnumerable<ITypedElement> focus, IEnumerable<ITypedElement> other)
-            => focus.Intersect(other, EqualityOperators.TypedElementEqualityComparer);
+        [TemporarilyChanged] // We cast all of them to scoped nodes for now. This will not be necessary once we define a clear Equality operator for IScopedNode
+        public static IEnumerable<IScopedNode> Intersect(this IEnumerable<IScopedNode> focus, IEnumerable<IScopedNode> other)
+            => focus.Intersect(other, EqualityOperators.TypedElementEqualityComparer).Select(ite => ite.ToScopedNode());
```
Comment (Kasdejong) [inline src/Hl7.Fhir.Base/FhirPath/Functions/CollectionOperators.cs]: ToScopedNodes()

File: src/Hl7.Fhir.Base/FhirPath/Functions/CollectionOperators.cs [hunk 7]
```diff
@@ -93,27 +98,28 @@ public static int LastIndexOf(this IEnumerable<ITypedElement> focus, ITypedEleme
...
         {
             if (char.IsUpper(name[0]))
             {
                 // If we are at a resource, we should match a path that is possibly not rooted in the resource
                 // (e.g. doing "name.family" on a Patient is equivalent to "Patient.name.family")   
                 // Also we do some poor polymorphism here: Resource.meta.lastUpdated is also allowed.
-                var baseClasses = new[] { "Resource", "DomainResource" };
-                if (element.InstanceType == name || baseClasses.Contains(name))
+#pragma warning disable CS0612 // Type or member is obsolete
+                if (element.InstanceType == name && element.Type.HasFlag(NodeType.Resource))
```
Comment (Kasdejong) [inline src/Hl7.Fhir.Base/FhirPath/Functions/CollectionOperators.cs]: keep the old code here

File: src/Hl7.Fhir.ElementModel.Shared.Tests/ElementNodeTests.cs [hunk 8]
```diff
@@ -63,6 +64,8 @@ private ElementNode createPatient()
         }
 
         [TestMethod]
+        [Ignore]
+        [TemporarilyChanged] // we will not implement this polymorphism properly for now, since we are moving to base soon.
```
Comment (Kasdejong) [inline src/Hl7.Fhir.ElementModel.Shared.Tests/ElementNodeTests.cs]: after reverting the code, this test should run again

File: src/Hl7.Fhir.ElementModel.Shared.Tests/ScopedNodeTests.cs [hunk 9]
```diff
@@ -271,17 +271,15 @@ static bool CCDATypeNameMapper(string typeName, out string canonical)
...
         }
         
         [TestMethod]
         public void Bundle_WithEntryWithoutFullUrl_ShouldNotThrow()
         {
             var bundle = new Bundle() { Type = Bundle.BundleType.Batch, Entry = [new Bundle.EntryComponent() { Resource = new Patient() }]}.ToTypedElement().ToScopedNode();
 
-            var enumerate = () => bundle.BundledResources();
-            enumerate.Should().NotThrow().Subject.Should().ContainSingle(c => c.FullUrl == null);
+            var enumerate = () => bundle.Children("entry");
```
Comment (Kasdejong) [inline src/Hl7.Fhir.ElementModel.Shared.Tests/ScopedNodeTests.cs]: BundledResources is fine here

File: src/Hl7.FhirPath.R4.Tests/PocoTests/FhirPathParallelTest.cs [hunk 10]
```diff
@@ -46,15 +46,14 @@ public async Tasks.Task TestSelectMethods()
...
         /// </summary>
-        public static async Tasks.Task MassiveParallelSelectsShouldBeCorrect(string testName, Func<ITypedElement, string, EvaluationContext, IEnumerable<ITypedElement>> selector)
+        public static async Tasks.Task MassiveParallelSelectsShouldBeCorrect(string testName, Func<IScopedNode, string, EvaluationContext, IEnumerable<ITypedElement>> selector)
         {
             var actual = new ConcurrentBag<(string canonical, ValueSet resource)>();
             var buffer = new BufferBlock<ValueSet>();
             var processor = new ActionBlock<ValueSet>(r =>
-                {
-                    var typedElement = r.ToTypedElement();
+                { ;
```
Comment (Kasdejong) [inline src/Hl7.FhirPath.R4.Tests/PocoTests/FhirPathParallelTest.cs]: remove semicolon?

File: src/Hl7.FhirPath.R4.Tests/PocoTests/FhirPathTest.cs [hunk 11]
```diff
@@ -35,6 +36,8 @@ public class FhirPathTest
         public TestContext TestContext { get; set; }
 
         [TestMethod]
+        [Ignore]
+        [TemporarilyChanged] // We will come back to polymorphism when we refactor against base
```
Comment (Kasdejong) [inline src/Hl7.FhirPath.R4.Tests/PocoTests/FhirPathTest.cs]: Should work again after reverting the code around polymorphism navigation

File: src/Hl7.FhirPath.R4.Tests/PocoTests/FhirPathTest.cs [hunk 12]
```diff
@@ -47,6 +50,8 @@ public void TestFhirPathPolymporphism()
         }
 
         [TestMethod]
+        [Ignore]
+        [TemporarilyChanged] // We will come back to polymorphism when we refactor against base
```
Comment (Kasdejong) [inline src/Hl7.FhirPath.R4.Tests/PocoTests/FhirPathTest.cs]: idem

File: src/Hl7.FhirPath.Tests/Functions/FunctionsTests.cs [hunk 13]
```diff
@@ -541,11 +541,11 @@ public void SingleScalarTest()
             {
                 iterations++;
 
-                return ElementNode.CreateList(iterations);
+                return ElementNode.CreateList(iterations).ToScopedNodes();
```
Comment (Kasdejong) [inline src/Hl7.FhirPath.Tests/Functions/FunctionsTests.cs]: use Integer here

File: src/Hl7.FhirPath.Tests/Tests/CastTests.cs [hunk 14]
```diff
@@ -173,6 +175,6 @@ public object Value
 
         public IElementDefinitionSummary Definition => null;
 
-        public IEnumerable<ITypedElement> Children(string name = null) => new ITypedElement[0];
+        public IEnumerable<ITypedElement> Children(string name = null) => Array.Empty<IScopedNode>();
```
Comment (Kasdejong) [inline src/Hl7.FhirPath.Tests/Tests/CastTests.cs]: collection expression, it kind of looks weird this way anyways!

**Review by ewoutkramer — APPROVED** [review]

**Comment by Kasdejong** [comment]
I'm assuming I can mark the conversations as resolved since you approved the PR @ewoutkramer?

**Pull request merged**